azext_iot/common/base_discovery.py

```diff
@@ -0,0 +1,342 @@
+                resource_list.extend(resource)
+        else:
+            try:
+                while True:
+                    resource_list.extend(resource_pager.advance_page())
+            except StopIteration:
+                pass
+
+        return resource_list
+
+    def get_policies(self, resource_name: str, rg: str) -> List:
+        """
+        Returns a list of all policies for a given resource in a given resource group.
+
+        :param resource_name: Resource Name
+        :type resource_name: str
+        :param rg: Resource Group
+        :type rg: str
+
+        :return: List of policies
+        :rtype: List
+        """
+        self._initialize_client()
+
+        policy_pager = self.client.list_keys(
+            **self._make_kwargs(resource_name=resource_name, resource_group_name=rg)
+        )
+        policy_list = []
+
+        if self.track2:
+            for policy in policy_pager.by_page():
+                policy_list.extend(policy)
+        else:
+            try:
+                while True:
+                    policy_list.extend(policy_pager.advance_page())
+            except StopIteration:
+                pass
+
+        return policy_list
+
+    def find_resource(self, resource_name: str, rg: str = None):
+        """
+        Returns the resource with the given resource_name.
+
+        If the resource group is not provided, will look through all resources within the
+        subscription and return first match. This functionality will only work for
+        resource types that require unique names within the subscription.
+
+        Raises CLIError if no resource is found.
+
+        :param resource_name: Resource Name
+        :type resource_name: str
+        :param rg: Resource Group
+        :type rg: str
+
+        :return: Resource
+        :rtype: dict representing self.resource_type
+        """
+        self._initialize_client()
+
+        if rg:
+            try:
+                return self.client.get(
+                    **self._make_kwargs(
+                        resource_name=resource_name, resource_group_name=rg
+                    )
+                )
+            except:  # pylint: disable=broad-except
+                raise CLIError(
+                    "Unable to find {}: {} in resource group: {}".format(
+                        self.resource_type, resource_name, rg
+                    )
+                )
+
+        resource_list = self.get_resources()
+
+        if resource_list:
+            target = next(
+                (resource for resource in resource_list if resource_name.lower() == resource.name.lower()),
+                None
+            )
+            if target:
+                return target
+
+        raise CLIError(
+            "Unable to find {}: {} in current subscription {}.".format(
+                self.resource_type, resource_name, self.sub_id
+            )
+        )
+
+    @abstractmethod
+    def _usable_policy(self, policy):
```

Comment:
Explicit type hints (even for return values) would be very helpful

Reply:
Added type hints